### ISSUE TITLE
[SKiDL BUG] tstamps/uuids not working correctly

### ISSUE THREAD

**Issue by cyberhuman** [devbisme/skidl#238]
**Describe the bug**
When working on a bigger project with auto-assigned component references, I noticed that when re-importing the netlist into the KiCAD's PCB editor, it doesn't associate the existing symbols with the symbols in the netlist. I tried [the Tags functionality](https://devbisme.github.io/skidl/#tags) but it didn't work (no change in behaviour), so I digged a bit and figured out it can't work in KiCAD8 (and maybe earlier versions) because it expects the following:

1. in the `(sheetpath .. (tstamps X))`, the components of X must be UUIDs.
2. there should be another `(tstamps Y)` as a sibling of the `(sheetpath)` node with the component UUID itself.

See the relevant source code here: https://gitlab.com/kicad/code/develop/-/blob/master/pcbnew/netlist_reader/kicad_netlist_reader.cpp#L442-473.
The sheetpath tstamps are KIID_PATH, which is a `/`-separated string of KIIDs, and the extra tstamps is a vector of KIIDs. A KIID is a wrapper for UUID which generates a random UUID if the constructor argument is not a valid UUID, which is what SKiDL currently writes.

**To Reproduce**

1. use the code:
```py
C = Part('Device', 'C', dest=TEMPLATE, footprint='Capacitor_SMD:C_0402_1005Metric')
C1 = C(value='10u', tag='C1')
C2 = C(value='4.7u', tag='C2')
```
2. generate the netlist
3. import the netlist into the KiCAD8's PCB editor
4. change the code (swap lines with C1 and C2):
```py
C = Part('Device', 'C', dest=TEMPLATE, footprint='Capacitor_SMD:C_0402_1005Metric')
C2 = C(value='4.7u', tag='C2')
C1 = C(value='10u', tag='C1')
```
5. generate the netlist
6. try to import the netlist into the KiCAD8's PCB editor and observe the behaviour:
6.1. when the "Link footprints using reference designators" option is used, the footprints are swapped.
6.2. when the "Link footprints using component tstamps (unique ids)" option is used, the existing footprints are removed and new footprints are added instead.

**Expected behavior**
When assigning a tag to a component, the component should get a valid stable UUID in the netlist.

**Additional context**
The following adhoc patch fixes the issue for me.
The `namespace_uuid` was generated using the `genuuid` command.
Note I don't really understand the semantics of the `(sheetpath (tstamps ...))` and the `(tstamps ...)`. The solution just generates stable UUIDs as expected by the PCB editor.
```diff
diff --git a/src/skidl/tools/kicad8/gen_netlist.py b/src/skidl/tools/kicad8/gen_netlist.py
index 7f878868..517c0754 100644
--- a/src/skidl/tools/kicad8/gen_netlist.py
+++ b/src/skidl/tools/kicad8/gen_netlist.py
@@ -8,12 +8,15 @@ Generate KiCad 5 netlist.
 
 import os.path
 import time
+import uuid
 
 from skidl.pckg_info import __version__
 from skidl.scriptinfo import scriptinfo
 from skidl.utilities import add_quotes, export_to_all
 
 
+namespace_uuid = uuid.UUID("7026fcc6-e1a0-409e-aaf4-6a17ea82654f")
+
 def gen_netlist_comp(part):
     """Generate the netlist text describing a component.
 
@@ -39,7 +42,11 @@ def gen_netlist_comp(part):
     # Embed the hierarchy along with a random integer into the sheetpath for each component.
     # This enables hierarchical selection in pcbnew.
     hierarchy = add_quotes("/" + part.hierarchical_name.replace(HIER_SEP, "/"))
-    tstamps = hierarchy
+    tstamps = "/".join([
+        str(uuid.uuid5(namespace_uuid, component))
+        for component in part.hierarchical_name.split(HIER_SEP)
+    ])
+    tstamps_uuid = uuid.uuid5(namespace_uuid, hierarchy)
 
     fields = ""
     for fld_name, fld_value in part.fields.items():
@@ -59,7 +66,7 @@ def gen_netlist_comp(part):
         + "      (footprint {footprint})\n"
         + "{fields}"
         + "      (libsource (lib {lib_filename}) (part {part_name}))\n"
-        + "      (sheetpath (names {hierarchy}) (tstamps {tstamps})))"
+        + "      (sheetpath (names {hierarchy}) (tstamps {tstamps})) (tstamps {tstamps_uuid}))"
     )
     txt = template.format(**locals())
     return txt
```

**Comment by devbisme** [comment]
Thanks for your bug report and your "adhoc" solution. I intend to incorporate it into the next release.

**Comment by devbisme** [comment]
I've placed your solution into the `development` branch (with a few cosmetic changes). If you have time, see if it works for you.

**Comment by shanemmattner** [comment]
For this issue I found that using `ref=xx` works to make the SKiDL output netlist use the desired references I use in Skidl code.  So for [this PR](https://github.com/devbisme/skidl/pull/243) the component outputs looking like:

```python
    C2 = Part('Device', 'C', value='10uF', footprint='Capacitor_SMD:C_0603_1608Metric', ref='C2', Sheetname='3v3_regulator', Sheetfile='power2.kicad_sch', ki_keywords='cap capacitor', ki_fp_filters='C_*')
```
Which imports back into Kicad nicely without changing anything.

I did not find that `tag` logic did anything.  Perhaps the [documentation](https://devbisme.github.io/skidl/#tags) needs updating @devbisme ?  Or am I misunderstanding the issue?

**Comment by devbisme** [comment]
Yes, explicitly setting the component references to match what's in the KiCad schematic will work to keep the SKiDL and the PCB consistent if you're importing the netlist using the references to match the components.

But, there is a problem that will probably arise. Suppose you are editing the SKiDL code and create a new component and assign it a reference of C4 and there is already a components somewhere in the code with that same reference. To prevent duplicate references, SKiDL will modify one of them to make them distinct. If SKiDL modifies the original reference, then that component won't match up with the corresponding schematic or PCB component. That's why KiCad has the option to import netlists using tstamps which are unique identifiers that get assigned when a component is added to a schematic and they stay constant even if the schematic component references are regenerated with new values. This maintains the correct association between the components in the schematic and the footprints in the PCB.

But SKiDL can't transparently assign unique, unchanging tstamps because all the components are regenerated each time the SKiDL script runs. So if you modify the script by adding new parts or by rearranging the code so the components are instantiated in a different order, then one or more components may be assigned a different tstamp and they won't match up with the KiCad components. The solution to this that I used in SKiDL is to have the user manually assign a tag to each component, subcircuit, and group and then generate the tstamps from these. A tstamp is computed using the tags from the hierarchical path that leads to a component, so the user only has to make sure that they do not assign identical tags to components within the same subcircuit.

But a problem still remains: the tstamps assigned by KiCad aren't going to be the same as those assigned by SKiDL. (I think KiCad may be using the date/time each component is instantiated in the schematic, while SKiDL uses the hierarchical tags.) So we will need a way to inject the tstamps generated by the SKiDL code into the KiCad schematic and PCB so that everybody remains consistent. Maybe that's just a simple search/replace from SKiDL to the KiCad schematic and PCB files, but I haven't looked at it in detail.

**Comment by cyberhuman** [comment]
Just my 5c. When my generated netlist has different uuids from what is in KiCAD (for example if I forget to add a tag and add it later, or change a tag), I usually import the netlist twice: first, using reference IDs (KiCAD then updates tstamps from the netlist), and then, using tstamps.

**Comment by devbisme** [comment]
Thanks for the tip, @cyberhuman! That works if you don't make too many changes to your SKiDL code. A problem arises if you add a component somewhere in the middle of the SKiDL code and then some references get re-assigned (if you haven't manually assigned them) and then you import the netlist into KiCad using references that no longer match up between the components. This doesn't happen if you use tags, but sometimes (as you said) you leave a tag out and then you might get into a mess. Maybe it's necessary to make `generate_netlist` issue a warning if you haven't manually assigned a tag?

**Comment by cyberhuman** [comment]
> Maybe it's necessary to make `generate_netlist` issue a warning if you haven't manually assigned a tag?

Sounds like a great idea!

**Comment by cyberhuman** [comment]
I have finally managed to test the fix for the original issue, and it works. Thank you!

**Comment by tapegoji** [comment]
I did a test on the fix. It works great. Thank you Dave!

**Comment by devbisme** [comment]
@cyberhuman , I modified `generate_netlist` so it issues a warning for each part that doesn't have a manually-assigned tag. It's currently in the `development` branch.
